proxy/grpcproxy/cluster.go

```diff
@@ -15,20 +15,99 @@
 package grpcproxy
 
 import (
+	"fmt"
+	"sync"
+
 	"github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/naming"
 	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
 
 	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+	gnaming "google.golang.org/grpc/naming"
 )
 
 type clusterProxy struct {
 	client *clientv3.Client
+	wa     gnaming.Watcher
+	donec  <-chan struct{}
+	umu    sync.RWMutex
+	umap   map[string]struct{}
 }
 
-func NewClusterProxy(c *clientv3.Client) pb.ClusterServer {
-	return &clusterProxy{
+// NewClusterProxy takes optional prefix to fetch grpc-proxy member endpoints.
+// The returned channel is closed when there is grpc-proxy endpoint registered
+// and the client's context is canceled so the 'register' loop returns.
+func NewClusterProxy(c *clientv3.Client, prefix string, addr string, ttl int) (pb.ClusterServer, <-chan struct{}, error) {
+	cp := &clusterProxy{
 		client: c,
 	}
+
+	var watcher gnaming.Watcher
+	var donec <-chan struct{}
+	var err error
+	if prefix != "" {
+		donec = register(c, prefix, addr, ttl)
+		gr := &naming.GRPCResolver{Client: c}
+		watcher, err = gr.Resolve(prefix)
+		if err != nil {
+			return nil, donec, fmt.Errorf("failed to resolve %q (%v)", prefix, err)
+		}
+		cp.wa = watcher
+		go cp.monitor()
+	}
+
+	return cp, donec, nil
+}
+
+func (cp *clusterProxy) monitor() {
+	rfunc := func() <-chan error {
+		errc := make(chan error, 1)
+		defer close(errc)
+		ups, err := cp.wa.Next()
+		if err != nil {
+			errc <- err
+			return errc
+		}
+
+		cp.umu.Lock()
+		if len(cp.umap) == 0 {
+			cp.umap = make(map[string]struct{}, len(ups))
+		}
+		for i := range ups {
+			switch ups[i].Op {
+			case gnaming.Add:
+				cp.umap[ups[i].Addr] = struct{}{}
+			case gnaming.Delete:
+				delete(cp.umap, ups[i].Addr)
+			}
+		}
+		cp.umu.Unlock()
+		return errc
+	}
+	for {
+		select {
+		case err := <-rfunc():
```

Comment:
this is sort of a "fake" `select`-- the select won't kick off until `rfunc()` returns